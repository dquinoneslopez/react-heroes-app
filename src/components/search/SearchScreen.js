import React, { useMemo } from 'react'
import queryString from "query-string";

import { useLocation } from 'react-router-dom';
import { heroes } from '../../data/heroes'
import { useForm } from '../../hooks/useform';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({history}) => {
    
    const location = useLocation();
    const { q = '' } = queryString.parse(location.search);
    
    const [formValues, handleInputChange, reset] = useForm({
        searchText: q
    });
    
    const { searchText } = formValues;
    
    const heroesFiltered = useMemo(() => getHeroesByName(q), [ q ]);
    
    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${searchText}`);
    }
    
    return (
        <div>
            <h1>Search Screen</h1>
            <hr/>
            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <form onSubmit={handleSearch}>
                        <input 
                            type="text" 
                            placeholder="find your hero"
                            className="form-control"
                            name="searchText"
                            value={searchText}
                            onChange={handleInputChange}
                            autoComplete="off"
                        />
                        <button
                            type="submit"
                            className="btn m-1 btn-outline-primary w-100"
                        >
                            Search
                        </button>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Results</h4>
                    <hr/>
                   
                    {
                        (q === '') && 
                        <div className="alert alert-info">
                            Search for a hero.
                        </div>
                    }
                    
                    {
                        (q !== '' && heroesFiltered.length === 0) && 
                        <div className="alert alert-danger">
                            There is no hero with the name '{q}'.
                        </div>
                    }
                    
                    {
                        heroesFiltered.map(hero => (
                            <HeroCard
                                key={hero.id}
                                {...hero}
                            />
                        ))
                    }
                </div>
            </div>
        </div>
    )
}
