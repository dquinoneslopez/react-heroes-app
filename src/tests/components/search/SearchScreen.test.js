import React from 'react';
import { mount } from 'enzyme'
import '@testing-library/jest-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';
import { MemoryRouter, Route } from 'react-router-dom';

describe('Pruebas en <SearchScreen/>', () => {

    const historyMock = {
        replace: jest.fn()
    }
    
    const wrapper = mount(
        <MemoryRouter initialEntries={['/search']}>
            <Route 
                path="/search"
                component={SearchScreen}
            />
        </MemoryRouter>
    );

    test('Debe mostrarse correctamente con valores por defecto', () => {
        
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search for a hero.');
        

    })
    
    test('Debe mostrar el héroe y el input con el valor del queryString', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route 
                    path="/search"
                    component={SearchScreen}
                />
            </MemoryRouter>
        );
        
        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();
        
    })
    
    test('Debe mostrar un error si no se encuentra el héroe', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=barman']}>
                <Route 
                    path="/search"
                    component={SearchScreen}
                />
            </MemoryRouter>
        );
        
        // console.log(wrapper.html())
        
        expect(wrapper.find('.alert-danger').exists()).toBe(true);
        expect(wrapper.find('.alert-danger').text().trim()).toEqual("There is no hero with the name 'barman'.");
        expect(wrapper).toMatchSnapshot();
        
    })

    test('Debe llamar el push del history', () => {
        
        const history = {
            push: jest.fn()
        }
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route 
                    path="/search"
                    component={() => <SearchScreen history={history} />}
                />
            </MemoryRouter>
        );
        
        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'batman'
            }
        });
        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });
        
        expect(history.push).toHaveBeenCalledWith(`?q=batman`)
        
    })
    

})