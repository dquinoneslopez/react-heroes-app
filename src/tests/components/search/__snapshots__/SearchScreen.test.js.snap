// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen/> Debe mostrar el héroe y el input con el valor del queryString 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-outline-primary w-100"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="col"
      >
        <div
          className="card ms-3 h-100"
          style={
            Object {
              "maxWidth": 540,
            }
          }
        >
          <div
            className="row no-gutters h-100"
          >
            <div
              className="col md-4"
            >
              <img
                alt="Batman"
                className="card-img h-100"
                src="./assets/heroes/dc-batman.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small>
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen/> Debe mostrar un error si no se encuentra el héroe 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="find your hero"
          type="text"
          value="barman"
        />
        <button
          className="btn m-1 btn-outline-primary w-100"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        There is no hero with the name '
        barman
        '.
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen/> Debe mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="find your hero"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-outline-primary w-100"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search for a hero.
      </div>
    </div>
  </div>
</div>
`;
